File: kata-others/kata-python/sorting_json_lines.py
import fileinput
import json
import re

'''
{"Bentley.G":{"balance":2134,"account_no":233831255}}
{"Alton.K":{"balance":9315,"account_no":203123613,"extra":{"balance":131}}}
{"Bancroft.M":{"balance": 233,"account_no":287655771101},"extra":{"balance":98}}
'''

def main():
    items = []
    for line in fileinput.input():
        jl = json.loads(line)
        
        name = next(iter(jl))
        data = jl[name]
        balance = (jl.get("extra") or data.get("extra", data)).get("balance")
        items.append((name, balance))

    for name, balance in sorted(items, key=lambda x: x[1]):
        print(f"{name}: {balance:,}")

if __name__ == '__main__':
    main()

File: kata-others/kata-javascript/SortingJsonLines.js
var readline = require('readline');
var rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
 });

 /*
{"Bentley.G":{"balance":2134,"account_no":233831255}}
{"Alton.K":{"balance":9315,"account_no":203123613,"extra":{"balance":131}}}
{"Bancroft.M":{"balance": 233,"account_no":287655771101},"extra":{"balance":98}}
*/
const list = []
rl.on('line', function(line){
    const jl = JSON.parse(line)
    const accountKey = Object.keys(jl).filter(k => k !== "extra")[0];
    const accountData = jl[accountKey];

    let balance;
    if (accountData && accountData.extra && accountData.extra.balance !== undefined) {
        balance = accountData.extra.balance; // inner extra
    } else if (jl.extra && jl.extra.balance !== undefined) {
        balance = jl.extra.balance; // outer extra
    } else if (accountData) {
        balance = accountData.balance; // normal balance
    }
    list.push({ name: accountKey, balance: balance })
});

rl.on('close', () => {
    list.sort((a, b) => a.balance - b.balance)
    for (const { name, balance } of list) {
        console.log(`${name}: ${balance.toLocaleString('en-US')}`)
    }
})